zcash_primitives/src/jubjub/mod.rs

```diff
@@ -427,6 +400,65 @@ impl JubjubBls12 {
 
         tmp_params
     }
+
+    fn find_group_hash<E: JubjubEngine>(
+        m: &[u8],
+        personalization: &[u8; 8],
+        params: &E::Params,
+    ) -> edwards::Point<E, PrimeOrder> {
+        let mut tag = m.to_vec();
+        let i = tag.len();
+        tag.push(0u8);
+
+        loop {
+            let gh = group_hash(&tag, personalization, params);
+
+            // We don't want to overflow and start reusing generators
+            assert!(tag[i] != u8::max_value());
+            tag[i] += 1;
+
+            if let Some(gh) = gh {
+                break gh;
+            }
+        }
+    }
+
+    /// Check for simple relations between the generators, that make finding collisions easy;
+    /// far worse than spec inconsistencies!
+    fn check_consistency_of_pedersen_hash_generators<E: JubjubEngine>(
+        tmp_params: &E::Params,
+        pedersen_hash_generators: &[edwards::Point<E, PrimeOrder>],
+    ) {
+        for (i, p1) in pedersen_hash_generators.iter().enumerate() {
+            if p1 == &edwards::Point::zero() {
+                panic!("Neutral element!");
+            }
+            for p2 in pedersen_hash_generators.iter().skip(i + 1) {
+                if p1 == p2 {
+                    panic!("Duplicate generator!");
+                }
+                if p1 == &p2.negate() {
+                    panic!("Inverse generator!");
+                }
+            }
+
+            // check for a generator being the sum of any other two
+            for (j, p2) in pedersen_hash_generators.iter().enumerate() {
```

Comment:
I experimented with trying to combine this loop with the one above, but wasn't able to get it to work properly. If this function were performance critical, we could first add all of the items to a HashSet, and then check if combinations of p1+p2 were in that set.